File: crates/kuben-platform/src/projection/informer.rs
//! Informers: one shared watch per kind, with backoff, label selectors and
//! `managedFields` stripped, feeding [`Projections`].

use std::{fmt::Debug, sync::Arc, time::Duration};

use futures::{StreamExt, TryStreamExt};
use k8s_openapi::api::core::v1::Pod;
use kube::{
    Api, Client, Resource, ResourceExt,
    api::{ApiResource, DynamicObject, GroupVersionKind},
    runtime::{WatchStreamExt, watcher},
};
use kuben_crd::{App, Environment, Project, labels};
use serde::de::DeserializeOwned;
use tokio_util::sync::CancellationToken;

use super::{AppView, CertificateView, EnvironmentView, PodView, ProjectView, Projections, RouteView};
use crate::{
    controller::gateway::TLS_SECRET_PREFIX,
    discovery::{CERT_MANAGER_GROUP, GATEWAY_GROUP},
    registry::ClusterRegistry,
};

/// Page size for the initial LIST. Keeps CPU bursts bounded on large clusters.
const PAGE_SIZE: u32 = 500;

/// How often an optional kind's API group is looked for until it appears.
const OPTIONAL_CHECK: Duration = Duration::from_mins(1);

/// How long the informers get for their first LIST before they are started
/// over. A request that hangs on an API server that has only just come up
/// would otherwise keep Kuben unready until the client's read timeout,
/// minutes later.
const FIRST_SYNC_DEADLINE: Duration = Duration::from_secs(45);

/// Run all informers for the primary cluster until cancelled.
pub async fn run(
    registry: ClusterRegistry,
    projections: Arc<Projections>,
    token: CancellationToken,
) -> anyhow::Result<()> {
    let client = registry.primary();
    let deadline = first_sync_deadline(projections.clone());
    tokio::select! {
        () = watch_pods(client.clone(), projections.clone(), token.child_token()) => Ok(()),
        () = watch_projects(client.clone(), projections.clone(), token.child_token()) => Ok(()),
        () = watch_environments(client.clone(), projections.clone(), token.child_token()) => Ok(()),
        () = watch_apps(client.clone(), projections.clone(), token.child_token()) => Ok(()),
        () = watch_routes(client.clone(), projections.clone(), token.child_token()) => Ok(()),
        () = watch_certificates(client, projections, token.child_token()) => Ok(()),
        e = deadline => Err(e),
        () = token.cancelled() => Ok(()),
    }
}

/// Resolves, with the informers still waiting, when the first LIST has not
/// completed within [`FIRST_SYNC_DEADLINE`]; never once it has.
async fn first_sync_deadline(projections: Arc<Projections>) -> anyhow::Error {
    tokio::select! {
        () = projections.wait_synced() => std::future::pending().await,
        () = tokio::time::sleep(FIRST_SYNC_DEADLINE) => anyhow::anyhow!(
            "no complete LIST within {}s for: {}",
            FIRST_SYNC_DEADLINE.as_secs(),
            projections.pending_kinds().join(", ")
        ),
    }
}

/// A generic watch loop. `on_event` receives every event; `Init*` events are
/// staged and swapped in at `InitDone` so readers never see a half-empty set.
async fn watch_kind<K, F>(api: Api<K>, cfg: watcher::Config, token: CancellationToken, mut on_event: F)
where
    K: Resource + Clone + DeserializeOwned + Debug + Send + 'static,
    K::DynamicType: Default,
    F: FnMut(Event<K>) + Send,
{
    let kind = K::kind(&K::DynamicType::default()).into_owned();
    watch_named(api, cfg, &kind, token, &mut on_event).await;
}

/// [`watch_kind`] for any object, named `kind` in the logs.
async fn watch_named<K, F>(
    api: Api<K>,
    cfg: watcher::Config,
    kind: &str,
    token: CancellationToken,
    on_event: &mut F,
) where
    K: Resource + Clone + DeserializeOwned + Debug + Send + 'static,
    F: FnMut(Event<K>) + Send,
{
    let stream = watcher(api, cfg)
        .default_backoff()
        .modify(|obj| obj.managed_fields_mut().clear())
        .map_ok(|ev| match ev {
            watcher::Event::Init => Event::Init,
            watcher::Event::InitApply(o) => Event::InitApply(o),
            watcher::Event::InitDone => Event::InitDone,
            watcher::Event::Apply(o) => Event::Apply(o),
            watcher::Event::Delete(o) => Event::Delete(o),
        });
    let mut stream = std::pin::pin!(stream.take_until(token.cancelled_owned()));
    while let Some(ev) = stream.next().await {
        match ev {
            Ok(ev) => on_event(ev),
            // The watcher lists again by itself, paced by the backoff. Ending
            // here would restart every informer, with a delay that grows on
            // each error: on a cluster that has just started, where the
            // first requests fail (a CRD not there yet), that kept Kuben
            // unready for minutes.
            Err(e) => tracing::warn!(kind = %kind, error = %e, "watch failed; retrying"),
        }
    }
}

/// Backend-agnostic watch event (mirrors `kube::runtime::watcher::Event`).
#[derive(Debug)]
pub enum Event<K> {
    Init,
    InitApply(K),
    InitDone,
    Apply(K),
    Delete(K),
}

async fn watch_pods(client: Client, projections: Arc<Projections>, token: CancellationToken) {
    let api = Api::<Pod>::all(client);
    let cfg = watcher::Config::default()
        .labels(labels::MANAGED_SELECTOR)
        .page_size(PAGE_SIZE);
    let mut staging: Vec<PodView> = Vec::new();
    watch_kind(api, cfg, token, move |ev| match ev {
        Event::Init => staging.clear(),
        Event::InitApply(p) => staging.push(PodView::from(&p)),
        Event::InitDone => {
            tracing::info!(pods = staging.len(), "pod informer synced");
            projections.replace_pods(std::mem::take(&mut staging));
        }
        Event::Apply(p) => projections.upsert_pod(PodView::from(&p)),
        Event::Delete(p) => {
            projections.remove_pod(&format!("{}/{}", p.namespace().unwrap_or_default(), p.name_any()));
        }
    })
    .await;
}

async fn watch_projects(client: Client, projections: Arc<Projections>, token: CancellationToken) {
    let api = Api::<Project>::all(client);
    let cfg = watcher::Config::default().page_size(PAGE_SIZE);
    let mut staging: Vec<ProjectView> = Vec::new();
    watch_kind(api, cfg, token, move |ev| match ev {
        Event::Init => staging.clear(),
        Event::InitApply(p) => staging.push(ProjectView::from(&p)),
        Event::InitDone => {
            tracing::info!(projects = staging.len(), "project informer synced");
            projections.replace_projects(std::mem::take(&mut staging));
        }
        Event::Apply(p) => projections.upsert_project(ProjectView::from(&p)),
        Event::Delete(p) => projections.remove_project(&p.name_any()),
    })
    .await;
}

async fn watch_environments(client: Client, projections: Arc<Projections>, token: CancellationToken) {
    let api = Api::<Environment>::all(client);
    let cfg = watcher::Config::default().page_size(PAGE_SIZE);
    let mut staging: Vec<EnvironmentView> = Vec::new();
    watch_kind(api, cfg, token, move |ev| match ev {
        Event::Init => staging.clear(),
        Event::InitApply(e) => staging.push(EnvironmentView::from(&e)),
        Event::InitDone => {
            tracing::info!(environments = staging.len(), "environment informer synced");
            projections.replace_environments(std::mem::take(&mut staging));
        }
        Event::Apply(e) => projections.upsert_environment(EnvironmentView::from(&e)),
        Event::Delete(e) => projections.remove_environment(&e.name_any()),
    })
    .await;
}

async fn watch_apps(client: Client, projections: Arc<Projections>, token: CancellationToken) {
    let api = Api::<App>::all(client);
    let cfg = watcher::Config::default().page_size(PAGE_SIZE);
    let mut staging: Vec<AppView> = Vec::new();
    watch_kind(api, cfg, token, move |ev| match ev {
        Event::Init => staging.clear(),
        Event::InitApply(a) => staging.push(AppView::from(&a)),
        Event::InitDone => {
            tracing::info!(apps = staging.len(), "app informer synced");
            projections.replace_apps(std::mem::take(&mut staging));
        }
        Event::Apply(a) => projections.upsert_app(AppView::from(&a)),
        Event::Delete(a) => {
            projections.remove_app(&format!("{}/{}", a.namespace().unwrap_or_default(), a.name_any()));
        }
    })
    .await;
}

/// Watch a kind that may not be installed (Gateway API, cert-manager): look
/// for its API group every minute and watch once it is served. Optional kinds
/// never hold back readiness.
async fn watch_optional<F>(
    client: Client,
    resource: ApiResource,
    cfg: watcher::Config,
    token: CancellationToken,
    mut on_event: F,
) where
    F: FnMut(Event<DynamicObject>) + Send,
{
    loop {
        match client.list_api_groups().await {
            Ok(groups) if groups.groups.iter().any(|g| g.name == resource.group) => break,
            Ok(_) => {}
            Err(e) => tracing::debug!(error = %e, group = %resource.group, "cannot list API groups"),
        }
        tokio::select! {
            () = token.cancelled() => return,
            () = tokio::time::sleep(OPTIONAL_CHECK) => {}
        }
    }
    let kind = resource.kind.clone();
    let api = Api::<DynamicObject>::all_with(client, &resource);
    watch_named(api, cfg, &kind, token, &mut on_event).await;
}

async fn watch_routes(client: Client, projections: Arc<Projections>, token: CancellationToken) {
    let gvk = GroupVersionKind::gvk(GATEWAY_GROUP, "v1", "HTTPRoute");
    let resource = ApiResource::from_gvk_with_plural(&gvk, "httproutes");
    let cfg = watcher::Config::default()
        .labels(labels::MANAGED_SELECTOR)
        .page_size(PAGE_SIZE);
    let mut staging: Vec<RouteView> = Vec::new();
    watch_optional(client, resource, cfg, token, move |ev| match ev {
        Event::Init => staging.clear(),
        Event::InitApply(r) => staging.push(RouteView::from(&r)),
        Event::InitDone => {
            tracing::info!(routes = staging.len(), "route informer synced");
            projections.replace_routes(std::mem::take(&mut staging));
        }
        Event::Apply(r) => projections.upsert_route(RouteView::from(&r)),
        Event::Delete(r) => {
            projections.remove_route(&format!("{}/{}", r.namespace().unwrap_or_default(), r.name_any()));
        }
    })
    .await;
}

/// Only the certificates Kuben's gateway orders; a cluster may hold many
/// others.
async fn watch_certificates(client: Client, projections: Arc<Projections>, token: CancellationToken) {
    let gvk = GroupVersionKind::gvk(CERT_MANAGER_GROUP, "v1", "Certificate");
    let resource = ApiResource::from_gvk_with_plural(&gvk, "certificates");
    let cfg = watcher::Config::default().page_size(PAGE_SIZE);
    let ours = |c: &DynamicObject| c.name_any().starts_with(TLS_SECRET_PREFIX);
    let mut staging: Vec<CertificateView> = Vec::new();
    watch_optional(client, resource, cfg, token, move |ev| match ev {
        Event::Init => staging.clear(),
        Event::InitApply(c) if ours(&c) => staging.push(CertificateView::from(&c)),
        Event::InitDone => projections.replace_certificates(std::mem::take(&mut staging)),
        Event::Apply(c) if ours(&c) => projections.upsert_certificate(CertificateView::from(&c)),
        Event::Delete(c) if ours(&c) => {
            projections.remove_certificate(&format!(
                "{}/{}",
                c.namespace().unwrap_or_default(),
                c.name_any()
            ));
        }
        Event::InitApply(_) | Event::Apply(_) | Event::Delete(_) => {}
    })
    .await;
}

File: internal/kube/projection/informer.go
package projection

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"slices"
	"strings"
	"time"

	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/meta"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"k8s.io/client-go/discovery"
	"k8s.io/client-go/dynamic/dynamicinformer"
	"k8s.io/client-go/informers"
	"k8s.io/client-go/tools/cache"

	"github.com/Teamtem-dev/kuben/api/v1alpha1"
	"github.com/Teamtem-dev/kuben/internal/kube/registry"
	"github.com/Teamtem-dev/kuben/internal/kube/render"
)

// The informers (projection/informer.rs): one shared watch per kind, with
// label selectors and managedFields stripped, feeding the Projections.
// client-go's reflector lists in pages of 500 (the Rust PAGE_SIZE) and
// re-lists with backoff by itself.

// API groups of the optional kinds.
const (
	GatewayGroup     = "gateway.networking.k8s.io"
	CertManagerGroup = "cert-manager.io"
)

// Timing of the informers.
type Timing struct {
	// FirstSyncDeadline is how long the informers get for their first LIST
	// before Run fails (and its supervisor starts them over). A request
	// that hangs on an API server that has only just come up would
	// otherwise keep Kuben unready until the client's read timeout,
	// minutes later.
	FirstSyncDeadline time.Duration
	// OptionalCheck is how often an optional kind's API group is looked
	// for until it appears.
	OptionalCheck time.Duration
	// Poll is how often a started informer is checked for its first LIST.
	Poll time.Duration
}

// DefaultTiming is the Rust constants.
func DefaultTiming() Timing {
	return Timing{FirstSyncDeadline: 45 * time.Second, OptionalCheck: time.Minute, Poll: 100 * time.Millisecond}
}

// Run runs every informer of cluster (the primary) into p until ctx ends,
// with the default timing.
func Run(ctx context.Context, cluster registry.Cluster, p *Projections, logger *slog.Logger) error {
	return RunWith(ctx, cluster, p, logger, DefaultTiming())
}

// kind is one watched kind: its informer and how its objects become views.
type kind struct {
	name     string
	informer cache.SharedIndexInformer
	synced   cache.InformerSynced
	// replace swaps in the views of every stored object.
	replace func(objects []any)
}

// RunWith is Run with the given timing. It fails when the required kinds
// (pods, projects, environments, apps) have not completed their first LIST
// within the deadline; optional kinds (Gateway API routes, cert-manager
// certificates) are watched once their API group is served and never hold
// back readiness.
func RunWith(ctx context.Context, cluster registry.Cluster, p *Projections, logger *slog.Logger, t Timing) error {
	ctx, cancel := context.WithCancel(ctx)
	typed := informers.NewSharedInformerFactoryWithOptions(cluster.Typed, 0,
		informers.WithTweakListOptions(func(o *metav1.ListOptions) { o.LabelSelector = v1alpha1.ManagedSelector }),
		informers.WithTransform(stripManagedFields))
	kuben := dynamicinformer.NewDynamicSharedInformerFactory(cluster.Dynamic, 0)
	managed := dynamicinformer.NewFilteredDynamicSharedInformerFactory(cluster.Dynamic, 0, metav1.NamespaceAll,
		func(o *metav1.ListOptions) { o.LabelSelector = v1alpha1.ManagedSelector })
	defer func() {
		cancel()
		typed.Shutdown()
		kuben.Shutdown()
		managed.Shutdown()
	}()

	required := []kind{
		podKind(typed.Core().V1().Pods().Informer(), p, logger),
		kubenKind(kuben, v1alpha1.ProjectResource, p.UpsertProject, p.RemoveProject, p.ReplaceProjects,
			ProjectViewOf, clusterKey, logger),
		kubenKind(kuben, v1alpha1.EnvironmentResource, p.UpsertEnvironment, p.RemoveEnvironment, p.ReplaceEnvironments,
			EnvironmentViewOf, clusterKey, logger),
		kubenKind(kuben, v1alpha1.AppResource, p.UpsertApp, p.RemoveApp, p.ReplaceApps,
			AppViewOf, namespacedKey, logger),
	}
	for _, k := range required {
		if k.synced == nil {
			return fmt.Errorf("%s informer: no event handler", k.name)
		}
	}
	typed.Start(ctx.Done())
	kuben.Start(ctx.Done())
	if err := firstSync(ctx, p, required, t); err != nil {
		return err
	}
	logger.Info("informers synced")
	optional := []optionalKind{
		{group: GatewayGroup, start: func() kind { return routeKind(managed, p, logger) }},
		{group: CertManagerGroup, start: func() kind { return certificateKind(kuben, p, logger) }},
	}
	watchOptional(ctx, cluster.Typed.Discovery(), optional, []func(<-chan struct{}){managed.Start, kuben.Start}, t, logger)
	return nil
}

// firstSync swaps in each required kind as soon as its first LIST is
// complete, and fails at the deadline naming those still listing.
func firstSync(ctx context.Context, p *Projections, kinds []kind, t Timing) error {
	deadline := time.NewTimer(t.FirstSyncDeadline)
	defer deadline.Stop()
	tick := time.NewTicker(t.Poll)
	defer tick.Stop()
	pending := slices.Clone(kinds)
	for {
		pending = slices.DeleteFunc(pending, func(k kind) bool {
			if !k.synced() {
				return false
			}
			k.replace(k.informer.GetStore().List())
			return true
		})
		if len(pending) == 0 {
			return nil
		}
		select {
		case <-ctx.Done():
			return nil
		case <-deadline.C:
			return fmt.Errorf("no complete LIST within %ds for: %s",
				int(t.FirstSyncDeadline/time.Second), strings.Join(p.PendingKinds(), ", "))
		case <-tick.C:
		}
	}
}

// optionalKind is a kind that may not be installed.
type optionalKind struct {
	group string
	start func() kind
	// running is set once the informer was created.
	running *kind
	synced  bool
}

// watchOptional looks for each optional kind's API group every
// OptionalCheck and watches the kind once it is served, until ctx ends.
func watchOptional(ctx context.Context, d discovery.DiscoveryInterface, kinds []optionalKind,
	starts []func(<-chan struct{}), t Timing, logger *slog.Logger,
) {
	for {
		waiting := false
		var groups []string
		for i := range kinds {
			k := &kinds[i]
			if k.running == nil {
				if groups == nil {
					groups = servedGroups(d, logger)
				}
				if slices.Contains(groups, k.group) {
					started := k.start()
					k.running = &started
					for _, start := range starts {
						start(ctx.Done())
					}
				}
			}
			if k.running != nil && !k.synced && k.running.synced != nil && k.running.synced() {
				k.running.replace(k.running.informer.GetStore().List())
				k.synced = true
				logger.Info("informer synced", "kind", k.running.name)
			}
			waiting = waiting || (k.running != nil && !k.synced)
		}
		wait := t.OptionalCheck
		if waiting {
			wait = t.Poll
		}
		select {
		case <-ctx.Done():
			return
		case <-time.After(wait):
		}
	}
}

func servedGroups(d discovery.DiscoveryInterface, logger *slog.Logger) []string {
	list, err := d.ServerGroups()
	if err != nil {
		logger.Debug("cannot list API groups", "error", err)
		return []string{}
	}
	groups := make([]string, 0, len(list.Groups))
	for _, g := range list.Groups {
		groups = append(groups, g.Name)
	}
	return groups
}

// stripManagedFields keeps memory low: managedFields are most of a small
// object and no view reads them.
func stripManagedFields(obj any) (any, error) {
	if m, err := meta.Accessor(obj); err == nil {
		m.SetManagedFields(nil)
	}
	return obj, nil
}

// objectKey is the namespace and name of a stored or deleted object.
func objectKey(obj any) (namespace, name string, ok bool) {
	if gone, isGone := obj.(cache.DeletedFinalStateUnknown); isGone {
		ns, n, err := cache.SplitMetaNamespaceKey(gone.Key)
		return ns, n, err == nil
	}
	m, err := meta.Accessor(obj)
	if err != nil {
		return "", "", false
	}
	return m.GetNamespace(), m.GetName(), true
}

func namespacedKey(namespace, name string) string { return namespace + "/" + name }

func clusterKey(_, name string) string { return name }

// handlers turns informer events into upserts and removes. Adds of the
// initial LIST are left to the replace at the first sync, so readers never
// see a half-filled set; later re-lists arrive as individual events.
func handlers[V any](view func(obj any) (V, bool), upsert func(V), remove func(string),
	key func(namespace, name string) string,
) cache.ResourceEventHandler {
	return cache.ResourceEventHandlerDetailedFuncs{
		AddFunc: func(obj any, initial bool) {
			if v, ok := view(obj); ok && !initial {
				upsert(v)
			}
		},
		UpdateFunc: func(_, obj any) {
			if v, ok := view(obj); ok {
				upsert(v)
			}
		},
		DeleteFunc: func(obj any) {
			if ns, name, ok := objectKey(obj); ok {
				remove(key(ns, name))
			}
		},
	}
}

func register[V any](name string, inf cache.SharedIndexInformer, view func(obj any) (V, bool), upsert func(V),
	remove func(string), replace func([]V), key func(namespace, name string) string, logger *slog.Logger,
) kind {
	k := kind{name: name, informer: inf, replace: func(objects []any) {
		views := make([]V, 0, len(objects))
		for _, o := range objects {
			if v, ok := view(o); ok {
				views = append(views, v)
			}
		}
		replace(views)
		logger.Info("informer listed", "kind", name, "objects", len(views))
	}}
	reg, err := inf.AddEventHandler(handlers(view, upsert, remove, key))
	if err != nil || reg == nil {
		logger.Error("cannot watch", "kind", name, "error", err)
		return k
	}
	k.synced = reg.HasSynced
	return k
}

func podKind(inf cache.SharedIndexInformer, p *Projections, logger *slog.Logger) kind {
	view := func(obj any) (PodView, bool) {
		pod, ok := obj.(*corev1.Pod)
		if !ok {
			return PodView{}, false
		}
		return PodViewOf(pod), true
	}
	return register("pods", inf, view, p.UpsertPod, p.RemovePod, p.ReplacePods, namespacedKey, logger)
}

// decode reads an unstructured kuben.dev object into its type, through
// the type's own JSON decoder (which applies the serde defaults).
func decode[T any](obj any) (T, error) {
	var out T
	u, ok := obj.(*unstructured.Unstructured)
	if !ok {
		return out, fmt.Errorf("unexpected %T", obj)
	}
	data, err := u.MarshalJSON()
	if err != nil {
		return out, err
	}
	if err := json.Unmarshal(data, &out); err != nil {
		return out, err
	}
	return out, nil
}

func kubenKind[T any, V any](f dynamicinformer.DynamicSharedInformerFactory, resource string,
	upsert func(V), remove func(string), replace func([]V), of func(*T) V,
	key func(namespace, name string) string, logger *slog.Logger,
) kind {
	gvr := v1alpha1.SchemeGroupVersion.WithResource(resource)
	inf := f.ForResource(gvr).Informer()
	if err := inf.SetTransform(stripManagedFields); err != nil {
		logger.Warn("cannot strip managedFields", "resource", resource, "error", err)
	}
	view := func(obj any) (V, bool) {
		t, err := decode[T](obj)
		if err != nil {
			// The Rust watcher reported an undecodable object and went on.
			logger.Warn("watch failed; skipping object", "resource", resource, "error", err)
			var zero V
			return zero, false
		}
		return of(&t), true
	}
	return register(resource, inf, view, upsert, remove, replace, key, logger)
}

func routeKind(f dynamicinformer.DynamicSharedInformerFactory, p *Projections, logger *slog.Logger) kind {
	inf := f.ForResource(schema.GroupVersionResource{Group: GatewayGroup, Version: "v1", Resource: "httproutes"}).Informer()
	if err := inf.SetTransform(stripManagedFields); err != nil {
		logger.Warn("cannot strip managedFields", "resource", "httproutes", "error", err)
	}
	view := func(obj any) (RouteView, bool) {
		u, ok := obj.(*unstructured.Unstructured)
		if !ok {
			return RouteView{}, false
		}
		return RouteViewOf(u), true
	}
	return register("httproutes", inf, view, p.UpsertRoute, p.RemoveRoute, p.ReplaceRoutes, namespacedKey, logger)
}

// certificateKind watches only the certificates Kuben's gateway orders; a
// cluster may hold many others.
func certificateKind(f dynamicinformer.DynamicSharedInformerFactory, p *Projections, logger *slog.Logger) kind {
	inf := f.ForResource(schema.GroupVersionResource{Group: CertManagerGroup, Version: "v1", Resource: "certificates"}).Informer()
	if err := inf.SetTransform(stripManagedFields); err != nil {
		logger.Warn("cannot strip managedFields", "resource", "certificates", "error", err)
	}
	view := func(obj any) (CertificateView, bool) {
		u, ok := obj.(*unstructured.Unstructured)
		if !ok || !strings.HasPrefix(u.GetName(), render.TLSSecretPrefix) {
			return CertificateView{}, false
		}
		return CertificateViewOf(u), true
	}
	remove := func(key string) {
		if _, name, ok := strings.Cut(key, "/"); ok && strings.HasPrefix(name, render.TLSSecretPrefix) {
			p.RemoveCertificate(key)
		}
	}
	return register("certificates", inf, view, p.UpsertCertificate, remove, p.ReplaceCertificates, namespacedKey, logger)
}
